pkg/vminetworkscontroller/vmi_controller.go

```diff
@@ -202,29 +198,42 @@ func (r *VirtualMachineInstanceReconciler) vmiNetworksClaimingIPAM(
 	return vmiNets, nil
 }
 
-func (r *VirtualMachineInstanceReconciler) Cleanup(vmiKey apitypes.NamespacedName) error {
-	ipamClaims := &ipamclaimsapi.IPAMClaimList{}
-	listOpts := []client.ListOption{
-		client.InNamespace(vmiKey.Namespace),
-		ownedByVMLabel(vmiKey.Name),
-	}
-	if err := r.Client.List(context.Background(), ipamClaims, listOpts...); err != nil {
-		return fmt.Errorf("could not get list of IPAMClaims owned by VM %q: %w", vmiKey.String(), err)
+func shouldCleanFinalizers(vmi *virtv1.VirtualMachineInstance, vm *virtv1.VirtualMachine) bool {
+	return (vm != nil && vmi == nil && vm.DeletionTimestamp != nil) ||
+		(vm == nil && (vmi == nil || (vmi.DeletionTimestamp != nil && len(vmi.Status.ActivePods) == 0)))
+}
```

Comment:
It may be a little more verbose but it will improve readability, is easy to forget about this in a few days, you can even wrap them at helper functions to make it  more clear
```suggestion
func shouldCleanFinalizers(vmi *virtv1.VirtualMachineInstance, vm *virtv1.VirtualMachine) bool {
        // Everything is gone
       if vm == nil && vmi == nil {
           return true
       } 
        
        // VMI is gone and VM is being deleted
        if vmi == nil && vm! == nil && vm.DeletionTimestamp != nil {
           return true
        } 
        
        // Standalone VMI being deleted and no virt-launcher pods
        if vm == nil && vmi != nil && vmi.DeletionTimestamp != nil && len(vmi.Status.ActivePods) == 0 {
            return true
        }
        
        return false
```